docs: add "snapshots" reference (#5363)
canonical/snapcraft: docs/reference/snapshots.rst

.. _reference-snapshots:

Snapshots
=========

A snap's system and user data can be excluded from `snapshots
<https://snapcraft.io/docs/snapshots>`_ by specifying exclusion patterns in an optional
metadata file called ``snapshots.yaml``. Such exclusions can be used to control snapshot
content and size.

.. important::

    The ``snap restore`` command replaces a snap's system and user data with the
    snapshot content, meaning that excluded files and directories will be lost.

``snapshots.yaml`` syntax
-------------------------

The ``snapshots.yaml`` configuration file starts with the ``exclude`` keyword followed
by a list of shell-style wildcard patterns to indicate which files or directories to
exclude.

These wildcard patterns must start with a system or user data `environment variable

> <https://snapcraft.io/docs/parts-environment-variables>`_. Only the asterisk wildcard is

Author: This link should get replaced with `reference-part-environment-variables` once it has been migrated.

supported.

The following example excludes cached files in ``.cache``, files beginning with
``excl-``, temporary files stored in ``/tmp``, and files manually stored in a
``large-files`` directory.

.. code-block:: yaml
    :caption: snapshots.yaml

    exclude:
      - $SNAP_COMMON/.cache
      - $SNAP_DATA/excl-*
      - $SNAP_USER_COMMON/tmp
      - $SNAP_USER_DATA/large-files/*-not-used.*

Including ``snapshots.yaml`` in a snap
--------------------------------------

> The ``snapshots.yaml`` file must be located within a snap's ``meta`` directory. This is
> typically done by `creating a part <https://snapcraft.io/docs/adding-parts>`_ that uses
> either the dump plugin or a build step override to copy ``snapshots.yaml`` from another
> directory.

Reviewer: I don't like the second example from the original doc with the `nil` plugin, so here's my suggestion:
```suggestion
The ``snapshots.yaml`` file must be located within a snap's ``meta`` directory. This is
typically done by `creating a part <https://snapcraft.io/docs/adding-parts>`_ that uses the dump plugin and the ``organize`` key to move ``snapshots.yaml`` into the ``meta`` directory.
```

The first example is short and sweet, is there a reason you kept the example snapshots.yaml content above but not an example of the part?

This is the example I like from the docs:
```yaml
parts:
  data-creation:
  # …
  snapshot-exclusions:
    source: config/
    plugin: dump
    organize:
      snapshots.yaml: meta/
```

Author: I cut the examples because I don't think they fit within the scope of reference documentation. They are very practical and useful, but those qualities are exactly why they should go in a how-to.

However, the example I didn't cut falls into the same category, so I'm going to rework it to convey the syntax in a way that is more appropriate for a reference doc.

Reviewer: looks good!

> directory.

Reviewer 2: I think it's serviceable. If you wanted to amp it up a bit inexpensively, you could link to [dump plugin](https://canonical-snapcraft.readthedocs-hosted.com/en/stable/common/craft-parts/explanation/dump_plugin.html) and [Overriding the default build](https://canonical-snapcraft.readthedocs-hosted.com/en/stable/common/craft-parts/how-to/override_build.html).

> directory.

Reviewer 2: Thanks!
